### PR TITLE
Add GHC 9.2 support

### PR DESCRIPTION
**Pull request by brandon-leapyear**
Related: #1475

### CODE REVIEW

**Review by akhesaCaro — CHANGES_REQUESTED** [review]
Thanks for your contribution,

Here some additional tasks that mustn't be forgotten :
- [ ] Be sure that the Nix can handle GHC 9.2 
- [ ] Add the parameter in the nix/README
- [ ] Add a check with Cabal on the CI (Only the stack check have been updated on the CI) - see https://github.com/haskell-servant/servant/blob/master/.github/workflows/master.yml

Are you okay with doing that?

Note : The PR (#1475) must be merged and this PR rebased before integrating it to the code base.

**Comment by brandon-leapyear** [comment]
I rebased and updated nix/README + github actions. I don't use Nix, so I won't be able to help with the Nix point, and I don't know what you mean by "Cookbook"

File: servant/src/Servant/Links.hs
Comment (brandon-leapyear) [inline servant/src/Servant/Links.hs:652]: This was added 2 weeks ago: https://github.com/haskell-servant/servant/commit/42b7d0eb9bc36e02fc4aa2656bcbde9aa540aaee

GHC 9.2 is stricter about this:

> GHC is stricter about checking for out-of-scope type variables on the right-hand sides of associated type family instances that are not bound on the left-hand side.

[source](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/9.2.1-notes.html)

**Comment by brandon-leapyear** [comment]
@ysangkok I changed GHC 9.2 to run with Cabal 3.6, but I'm still getting errors. I don't use Cabal myself, can you help fix that?

**Comment by akhesaCaro** [comment]
Okay I will try on my side and keep you posted.

**Comment by gdeest** [comment]
I am not entirely making sense of the Cabal failure ; it seems to me that Cabal commits to a bound-to-fail plan because the constraints it tries to satisfy do not match pre-installed versions, but I am not entirely sure.

I am slightly concerned with the amount of `allow-newer` stanzas we have in the `cabal.project` files:

```
-- https://github.com/chordify/haskell-servant-pagination/pull/12
allow-newer: servant-pagination-2.2.2:servant
allow-newer: servant-pagination-2.2.2:servant-server

allow-newer: servant-js:servant

-- ghc 9
allow-newer: tdigest:base

-- ghc 9.2
allow-newer: basement:base
allow-newer: binary:bytestring
allow-newer: generics-sop:base
allow-newer: generics-sop:ghc-prim
allow-newer: generics-sop:template-haskell
allow-newer: http-media:base
allow-newer: lucid:base
allow-newer: memory:base
allow-newer: openssl-streams:bytestring
allow-newer: servant-js:base
allow-newer: servant-js:base-compat
allow-newer: servant-js:lens
allow-newer: servant-multipart:bytestring
allow-newer: servant-swagger:base-compat
allow-newer: singleton-bool:base
allow-newer: sop-core:base
allow-newer: swagger:base
allow-newer: swagger2:base
allow-newer: swagger2:base-compat-batteries
allow-newer: swagger2:bytestring
allow-newer: swagger2:lens
allow-newer: swagger2:template-haskell
allow-newer: swagger2:time
```

I am guessing we don't really have a choice because these libraries do not have 9.2-compatible releases yet >

**Comment by gdeest** [comment]
I have managed to get past Cabal configure, but `basement` fails to build with GHC 9.2. We'll have to exclude some things from the 9.2 build.

**Comment by tchoutri** [comment]
Yes, the Hackage & Stackage folks have had some issues with [basement](https://github.com/haskell-infra/hackage-trustees/issues/330) and [foundation](https://github.com/commercialhaskell/stackage/issues/6400#issuecomment-1019001230), looks like the 9.2 support for these libraries will come one day™

File: servant/src/Servant/Links.hs
Comment (josephcsible) [reply servant/src/Servant/Links.hs:652]: aab7e0d added another instance of this problem: https://github.com/haskell-servant/servant/blob/7a770b5a1e0fc250c5f19c0629329e28b69151c0/servant-server/src/Servant/Server/Internal.hs#L824
`HasServer` needs to become `(HasServer :: * -> [*] -> Constraint)` (and `Constraint` needs to be added to the import list).

**Comment by brandon-leapyear** [comment]
@gdeest in the future, can you let me know before you push changes to my fork? Thanks!

File: servant-server/src/Servant/Server/Internal.hs
Comment (josephcsible) [inline servant-server/src/Servant/Server/Internal.hs:826]: Shouldn't that be `Type -> [Type] -> Constraint`?

**Comment by ysangkok** [comment]
I think it is a bad idea to make the default `stack.yaml` on the master branch refer to unreleased versions of libraries (e.g. `lens`, `foundation`, etc). I see how it makes sense to do this for development purposes, though. But if this branch isn't supposed to be merged, maybe it would be clearer to mark it "Draft"? After dependencies are fixed or removed, it could be marked ready for review again.

File: stack.yaml
Comment (ysangkok) [inline stack.yaml:30]: This can be removed since a fixed lens version was released.

**Review by ysangkok — CHANGES_REQUESTED** [review]

File: stack.yaml
Comment (ysangkok) [inline stack.yaml:26]: This should be unnecessary since https://github.com/yesodweb/wai/commit/493610faeb7c2dedd28ecb43e6a6adfb6e0f4eac was released.

**Comment by brandon-leapyear** [comment]
I intended for the PR to be merged now. I wasn't sure how Stack is used in this project, e.g. if the standard build workflow is via Cabal, and the stack config is just there for reference. I've moved the Stack stuff into a new `stack-ghc-9.2.yaml` config.

I would prefer to not have a PR that I maintain open for an indefinite period of time. If you'd prefer to not merge any aspect of this PR at this time, feel free to close.

File: stack-ghc-9.2.yaml
Comment (ysangkok) [inline stack-ghc-9.2.yaml:39]: These lines are now unnecessary since http-streams 0.8.9.6 have been released containing aesiniath/http-streams/pull/126 (which you submitted)

File: stack-ghc-9.2.yaml
Comment (ysangkok) [inline stack-ghc-9.2.yaml:21]: These lines are now unnecessary since swagger2 version 2.8 has been released.

File: cabal.project
Comment (ysangkok) [inline cabal.project:103]: These lines are now unnecessary because swagger2 version 2.8 has been released with adjusted bounds.

File: cabal.project
Comment (ysangkok) [inline cabal.project:95]: This line is unnecessary since a revision was published: https://hackage.haskell.org/package/singleton-bool-0.1.6/revisions/

File: cabal.project
Comment (ysangkok) [inline cabal.project:96]: This line is unneceessary since sop-core 0.5.0.2 was released: https://hackage.haskell.org/package/sop-core

File: cabal.project
Comment (ysangkok) [inline cabal.project:78]: These lines are unnecessary since generics-sop 0.5.1.2 has been released: https://hackage.haskell.org/package/generics-sop

File: cabal.project
Comment (ysangkok) [inline cabal.project:97]: Why was this added? The bounds are `>=4.6 && <5.0` which is wide enough.

**Review by ysangkok — CHANGES_REQUESTED** [review]

File: cabal.project
Comment (brandon-leapyear) [reply cabal.project:97]: 🤷 it was needed when I opened this PR 3 months ago

File: cabal.project
Comment (ysangkok) [inline cabal.project:77]: These have been updated and can be removed: binary, http-media, lucid

File: cabal.project
Comment (ysangkok) [inline cabal.project:91]: servant-swagger has been updated and these lines can be removed

**Review by ysangkok — CHANGES_REQUESTED** [review]

File: servant-server/src/Servant/Server/Internal.hs
Comment (shlevy) [inline servant-server/src/Servant/Server/Internal.hs:826]: Can we get a quick patch release with just this change included? With this I think GHC 9.2.1 compilation will succeed if allow-newer is set and fixed versions of cryptonite, memory, and basement are included in the cabal.project.

Comment (ysangkok) [reply servant-server/src/Servant/Server/Internal.hs:826]: It was already merged as part of

* haskell-servant/servant/pull/1525

File: cabal.project
Comment (ysangkok) [inline cabal.project:79]: These lines can be removed since [servant-js was revised](https://hackage.haskell.org/package/servant-js-0.9.4.2/revisions/).

File: cabal.project
Comment (ysangkok) [inline cabal.project:80]: * https://github.com/haskell-servant/servant-multipart/pull/59

File: cabal.project
Comment (ysangkok) [inline cabal.project:76]: * snapframework/openssl-streams/pull/6

File: cabal.project
Comment (ysangkok) [reply cabal.project:80]: @brandon-leapyear These ones for servant-multipart and the one for openssl-streams can be removed now. If you rebase, the diff will also get smaller now because allow-newers were removed on master.

**Comment by brandon-leapyear** [comment]
@ysangkok can we get this merged anytime soon? I'd prefer to not be maintaining this long-term

**Comment by ysangkok** [comment]
@brandon-leapyear I am not a project maintainer, so it is not my decision to make. But I have submitted #1555 which achieves GHC 9.2 compatibility for a subset of packages that work on GHC 9.2, and marks the others as unbuildable. I prefer this approach due to how it doesn't legitimize unofficial forks of dependencies (as [previously mentioned](https://github.com/haskell-servant/servant/pull/1482#issuecomment-1028160110)).

**Comment by brandon-leapyear** [comment]
Fine by me. Can you link the same issue in your PR?

**Comment by maksbotan** [comment]
@ysangkok @brandon-leapyear please check if this PR is still relevant after merge of #1555.

And thanks for your work!

**Pull request closed without merging**
